doc/gontributions.1.md
% gontributions(1) Version 0.7 | contributions saver and lister

NAME
====

**gontributions** — OSS contributions saver and lister

SYNOPSIS
========

| **gontributions** \[OPTIONS] \[SUBCOMMAND]

DESCRIPTION
===========

This program lists your open source contributions. They get printed on your terminal when running the program. Additionally an HTML page gets created. There are some templates in the templates directory. Feel free to add your own and create a pull request to this repository.

Sometimes upstream repositories disappear. And with it your contributions. Initially I wrote this program as a way to back up all the repositories I contributed to and get a list of my contributions. Either because I was proud of them, or because I might wanted to look at them again. In any case, I just like backups. From there the program grew, to create a small overview in HTML format so you could show the projects you contribute to to others.

Options
-------

\--config

:   Set which config file to use

\--template

: Set which template to use. If it contains a dot it will be treated as a path to a user defined template. No dot means using an internal template

\--output

: Define name of the generated HTMl file

\--no-pull

: Don't update VCS repositories


SUBCOMMANDS
=====

*exconf*

: Show an example configuration file

*list-templates*

: List all built-in templates

*show-template*

: Display the content of a template

CONFIGURATION
===========

```
gontributions exconf > example.json
$EDITOR example.json
```

BUGS
====

See GitHub Issues: <https://github.com/jubalh/gontributions/issues>

AUTHOR
======

gontributions is written by by Michael Vetter <jubalh@iodoru.org>.


CHANGELOG.md
# Changelog

## [0.7] 2023-01-11
* Add manpage
* Document requirements
* Update tests where test cases are not reachable anymore
* Minor code cleanups

## [0.6] 2018-04-11
* Add test for binaries needed for debian
* Add scanning of Debian testing changelog
* Add Bazaar support
* Add Mercurial test
* Increase Wiki test coverage
* Expand documentation
* Add possibility to have tags for projects
* Add summary to detailed template
* Only create repo directories when binary exists
* Add Mercurial support
* Add USAGE.md with MediaWiki
* Fixed issue with cloning repos named 'git@somehwere.com:repo.git'
* Show help if standard config file is not loadable
* Add no-pull switch
* Add date to templates
* Check if binaries exist before using them

## [0.5.2] 2016-07-22
* Use colors in PrintInfo function
* Update to newest cli version
* Log if we cant flush the file
* Add mediawiki test

## [0.5] 2016-02-12
* Add link to example report
* No empty description in default template
* Fix text in detailed template

## [0.4] 2016-02-12
* OBS: Count commits from multiple changes files
* Provide default templates in bindata asset
* Add FAQ and Contribute to README

## [0.3] 2016-01-31
* Add basic OpenBuildService support
* Add update function for OBS
* And example obs configuration
* Make templates valid HTML5

## [0.2] 2016-01-25
First version


gontributions.go
//go:generate go-bindata -pkg main -o default-templates-bindata.go templates/
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"html/template"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"

	"gopkg.in/urfave/cli.v1"

	"github.com/jubalh/gontributions/gontrib"
	"github.com/jubalh/gontributions/util"
	"github.com/jubalh/gontributions/vcs/mediawiki"
	"github.com/jubalh/gontributions/vcs/obs"
)

// Summary hold the overall nr of contributions and projects
// we do it like this so we can calculate here, in templates its not supported
type Summary struct {
	TotalContributions int
	ProjectCount       int
}

// TemplateFill is there so it can easily be extended without breaking old templates/layouts
type TemplateFill struct {
	Contributions []gontrib.Contribution
	Summary       Summary
}

const (
	templateFolderName = "templates"
	templatesFolderEnv = "GONTRIB_TEMPLATES_PTH"
)

// loadConfig loads a json configuration from filename
// and creates a Configuration from it.
func loadConfig(filename string) (gontribs gontrib.Configuration, err error) {
	file, err := os.OpenFile(filename, os.O_RDONLY, 0660)
	if err != nil {
		return
	}

	decoder := json.NewDecoder(file)
	err = decoder.Decode(&gontribs)
	return
}

// putdate() returns the current date.
// Meant for the template.
func putdate() string {
	return time.Now().Local().Format("2006-01-02")
}

// fillTemplate puts the information of a Contribution
// into a template.
func fillTemplate(contributions []gontrib.Contribution, tempContent string, writer io.Writer) {
	tf := TemplateFill{Contributions: contributions}

	for _, contribution := range contributions {
		tf.Summary.TotalContributions += contribution.Count
	}
	tf.Summary.ProjectCount = len(contributions)

	funcMap := template.FuncMap{
		"putdate": putdate,
	}

	t, err := template.New("string-template").Funcs(funcMap).Parse(tempContent)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	err = t.Execute(writer, tf)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

// Main will set and parse the cli options.
func main() {
	app := cli.NewApp()

	app.Name = "gontributions"
	app.Usage = "contributions lister"
	app.Author = "Michael Vetter"
	app.Version = "v0.5.2"
	app.Email = "jubalh@openmailbox.org"

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  "config",
			Value: "gontrib.json",
			Usage: "Set which config file to use",
		},
		cli.StringFlag{
			Name:  "template",
			Value: "default.html",
			Usage: "Set which template to use",
		},
		cli.StringFlag{
			Name:  "output",
			Value: "output.html",
			Usage: "Define name of the generated HTMl file",
		},
		cli.BoolFlag{
			Name:  "no-pull",
			Usage: "Don't update VCS repositories",
		},
	}
	app.Commands = []cli.Command{
		{
			Name:   "exconf",
			Usage:  "Show an example configuration file",
			Action: cmdExconf,
		},
	}

	app.Action = run

	err := app.Run(os.Args)
	if err != nil {
		fmt.Println(err.Error())
	}
}

func getTemplateOrExitError(ctx *cli.Context) (string, error) {
	// Get users template selection
	templateName := ctx.GlobalString("template")

	// Get Template as string
	templatesPath := os.Getenv(templatesFolderEnv)
	if templatesPath == "" {
		// Use asset
		data, err := Asset(filepath.Join(templateFolderName, templateName))
		if err != nil {
			return "", cli.NewExitError(err.Error(), 1)
		}
		return string(data), nil
	} else {
		// Use template from user defined folder
		absoluteTemplatePath := filepath.Join(templatesPath, templateName)
		if !util.FileExists(absoluteTemplatePath) {
			s := fmt.Sprintf("Template file %s does not exist\n", absoluteTemplatePath)
			return "", cli.NewExitError(s, 1)
		}
		data, err := ioutil.ReadFile(absoluteTemplatePath)
		if err != nil {
			return "", cli.NewExitError(err.Error(), 1)
		}
		return string(data), nil
	}
}

// notifyOnErrors; in case of a pull, update or other error regarding the checkout and
// processing of contributions; will print a message and the path to the error log
// otherwise if will remove any existing 'errors.log' (old errors from last run).
func notifyOnErrors() {
	errorfile, err := os.Open("errors.log")
	if err != nil {
		fmt.Println(err.Error())
		return
	}
	defer errorfile.Close()
	fi, err := errorfile.Stat()
	if err != nil {
		fmt.Println(err.Error())
		return
	}
	if fi.Size() > 0 {
		util.PrintInfoF(nil, "Some contributions could not be checked. See: errors.log", util.PI_ERROR)
	} else {
		os.Remove("errors.log")
	}
}

// Run will handle the functionallity.
func run(ctx *cli.Context) error {
	// Load specified json configuration file
	configPath := ctx.GlobalString("config")
	configuration, err := loadConfig(configPath)
	if err != nil {
		// if config cant be loaded because default one is used
		// (set in StringFlag) and is not available, then show the usage.
		if !ctx.IsSet("config") {
			cli.ShowAppHelp(ctx)
			return nil
		}
		return cli.NewExitError(err.Error(), 1)
	}

	// get template
	templateData, err := getTemplateOrExitError(ctx)
	if err != nil {
		return err
	}

	gontrib.PullSources = !ctx.GlobalBool("no-pull")

	// scan
	contributions, err := gontrib.ScanContributions(configuration)
	if err != nil {
		util.PrintInfo(nil, err.Error(), util.PI_ERROR)
		return cli.NewExitError(err.Error(), 1)
	}

	// define output
	outputPath := ctx.GlobalString("output")
	f, err := os.Create(outputPath)
	if err != nil {
		return cli.NewExitError(err.Error(), 1)
	}
	defer f.Close()

	writer := bufio.NewWriter(f)
	fillTemplate(contributions, templateData, writer)
	if err := writer.Flush(); err != nil {
		return cli.NewExitError(err.Error(), 1)
	}

	util.PrintInfoF(nil, "\nReport saved in: %s", util.PI_INFO, outputPath)

	notifyOnErrors()

	return nil
}

// Create an example configuration file which the user can
// adapt to his own needs.
func cmdExconf(c *cli.Context) error {
	configuration := gontrib.Configuration{
		Emails: []string{"jubalh@openmailbox.org", "g.bluehut@gmail.com"},
		Projects: []gontrib.Project{
			{Name: "nudoku", Description: "Ncurses sudoku game", Gitrepos: []string{"https://github.com/jubalh/nudoku"}, Tags: []string{"C", "Terminal", "Game"}},
			{Name: "profanity", Description: "Ncurses based XMPP client", URL: "http://profanity.im/", Gitrepos: []string{"https://github.com/boothj5/profanity"}, Tags: []string{"C", "XMPP", "Terminal"}},
			{Name: "Funtoo", Description: "Linux distribution", URL: "http://funtoo.org/", Gitrepos: []string{"https://github.com/funtoo/ego", "https://github.com/funtoo/metro"}, MediaWikis: []mediawiki.MediaWiki{{BaseUrl: "http://funtoo.org", User: "jubalh"}}},
			{Name: "openSUSE", Description: "Linux distribution", URL: "http://opensuse.org/", Obs: []obs.OpenBuildService{{Apiurl: "https://api.opensuse.org", Repo: "utilities/vifm"}}},
			{Name: "prosody", Description: "A modern XMPP communication server", URL: "http://prosody.im/", Hgrepos: []string{"https://hg.prosody.im/prosody-modules"}, Tags: []string{"Lua", "XMPP"}},
		},
	}

	text, err := json.MarshalIndent(configuration, "", "    ")
	if err != nil {
		return cli.NewExitError(err.Error(), 1)
	}

	fmt.Println(string(text))
	return nil
}
